### ISSUE TITLE
Switch expressions currently disallow trailing comma

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#2098]
Update: LDM (1/9) decided to allow such trailing comma.

-----

@jcouv commented on [Thu Dec 20 2018](https://github.com/dotnet/roslyn/issues/31962)

Trailing commas are currently disallowed, which is inconvenient when copy/pasting a line.

```C#
public class C 
{
    public void M() 
    {
        _ = 3 switch 
        {
                2 => 1,
                _ => 2, // error
        };
    }
}
```
[Example](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRIMaOQSwG4SoAOsMECAdsgD4CwAUAAIDMABIwEysDCrA3g1ZD2bRgBZWAWQAUASn6DhSgPqsAvKzbIA7jhgYAFq0VKhA+qcumuagHysAjABoTVy6rusOT1gHpfrOQIAPYIrkoAvgDcrhEMEUA===)

In contrast, we allow trailing commas in array initializers.

---

@CyrusNajmabadi commented on [Thu Dec 20 2018](https://github.com/dotnet/roslyn/issues/31962#issuecomment-449084019)

A big :+1: on this.  For anything that is commonly multi-line, not allowing the trailing comma is often a PITA.  Thanks!

---

@gafter commented on [Thu Dec 20 2018](https://github.com/dotnet/roslyn/issues/31962#issuecomment-449137843)

I advocated for this but the LDM shot it down.  I'll be happy to bring it back up based on "experience".

---

@CyrusNajmabadi commented on [Thu Dec 20 2018](https://github.com/dotnet/roslyn/issues/31962#issuecomment-449139601)

That would be great.  Thanks @gafter . 

---

@alrz commented on [Thu Dec 20 2018](https://github.com/dotnet/roslyn/issues/31962#issuecomment-449171298)

Hopefully also for property patterns which is an unordered list. 👍 

---

@sharwell commented on [Thu Dec 20 2018](https://github.com/dotnet/roslyn/issues/31962#issuecomment-449178605)

@gafter This will be super obvious to StyleCop Analyzers users if it turns out to be a "special" list (i.e. a list that does not allow a trailing comma). There's [a rule](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1413.md) for that.

**Comment by DavidArno** [comment]
> Trailing commas are currently disallowed

Good. Long may that continue to be the case too.

**Comment by YairHalberstadt** [comment]
A switch expression cannot ever get past the last case can it? Otherwise we would end up with an expression which doesn't always have a value.

In which case I think having a trailing comma is slightly misleading. It implies it's normal to add something after the last case, but actually you'd want to add it before the last case...

**Comment by alrz** [comment]
>it cannot be used with parameter lists

I think we should notice that's true almost always when there's parentheses rather than braces.

Adding a parameter changes the overload, adding a tuple element changes the type.

But adding a switch arm, property pattern, property initializer or enum member doesn't change anything about the type.

**Comment by alrz** [comment]
>A switch expression cannot ever get past the last case can it?

Switch expression doesn't necessarily ends with `_` case or alike. 
n an exhaustive switch expression (maybe on an enum without FlagsAttribute) you don't necessarily have a "default" arm, and the order wouldn't be important in these cases.

**Comment by qrli** [comment]
this is why i'd prefer leading symbol (like `case` or F#'s `|`) only than trailing delimiter only, so that no need to debate on such thing which is minor but important in feeling to some.

**Comment by CyrusNajmabadi** [comment]
This has nothing to do with parens/braces/etc.  This has to do with common ways of writing a list of elements and the continued annoyance around editing said lists.  When you have multiple lines containing these members, it is an extra annoyance to have to have the last line always be different from the rest.  It adds extra steps for the users for no extra benefit.

It's worth noting that other languages (like go) sidestep this entirely by actually making the trailing comma mandatory for similar situations.  And common linters for other languages (like TypeScript) require you to put these in for consistency.

--

Basically, no one is helped by *disallowing* this comma.  Whereas there are several patterns and editing scenarios that are helped by *allowing* it.  These patterns match the existing cases in the language where this is already allowed, and so we're also more consistent if this is allowed here.

**Comment by CyrusNajmabadi** [comment]
> It's a great shame then that the team changed the syntax from using parentheses to braces then...

Even with parens, this should be allowed.  It will be common to write this as:

```c#
expr switch
{
    p1 => v1,
    p2 => v2,
}
```

Adding a new case should be as simple as duplicating the last line.  The same as how simple it is to duplicate a line in any of our constructs today that allow trailing commas (like initializers). i.e. i can write this:

```c#
        new A
        {
            B = 1,
        };
```

And easily just add more items by copy/pasting, without having to go edit previous initializers as well.  This is a common editing improvement, recognized among tons of languages, and which should be allowed in any circumstance *where there is not risk of ambiguity*.  i.e. if i do this 

```c#
new X<
   int,
   //...
   string,
>

// or similar cases with tuples
```

there is a visual ambiguity here because it can lead people to think there may be an unstated item there that may be inferred or may otherwise have meaning.  In visual cases like that, or cases where the language might actually want a feature there in a future (like 'inferred type arguments'), then this shoudl not be allowed.  However, for completely safe areas, it should practically always be ok.

**Comment by DavidArno** [comment]
> This has nothing to do with parens/braces/etc. This has to do with common ways of writing a list of elements and the continued annoyance around editing said lists

Then fix the *editor*, rather than bodging the language...

**Comment by CyrusNajmabadi** [comment]
> Then fix the editor,

There isn't one editor to fix.  You've simply pushed the problem to everyone, instead of fixing at the natural location.

> rather than bodging the language...

It doesn't do anything negative to the language.  The language already supports this construct in many places *precisely* because it makes typing more pleasant.  There was no need to have ever supported this in hte past for our other constructs.  But we did, for this *exact* case.  The language is 'bodged' (i'm going to assume i know what that means) when new constructs don't behave consistently like these other ones do just because you want to be a zealot.

**Comment by sharwell** [comment]
Failure to include a comma at the end of a multi-line construct for which additional lines are likely to be added in the future is an unnecessary maintenance cost. When adding a new item at the end of a list, the following problems occur if the previous item did not have a trailing comma:

1. The pull request making the change will require reviewers look at a line which could have been avoided
2. After the previous step, **git blame** will produce incorrect results, requiring users to step back in the blame to find the information which should have been available from the start.

Some users may not agree with this, but there is a large body of C# developers which count on it continuing.

**Comment by alrz** [comment]
>Then fix the editor, rather than bodging the language

Same argument applies to `$@` and `$@`, 8.0 supports both. bodged.

**Comment by YairHalberstadt** [comment]
Can you give a scenario where you are likely to add something to the end of a switch expression between one commit and the other?

**Comment by alrz** [comment]
```diff
enum E
{
   Case1,
   Case2,
+  Case3,
}

var x = e switch
{
   Case1 => 1,
   Case2 => 2,
+  Case3 => 3,
}
```
Assuming enum without Flags is considered exhaustive.

**Comment by alrz** [comment]
Relates to  "sealed enums" (https://github.com/dotnet/csharplang/issues/782)

**Comment by jcouv** [comment]
Allowing trailing commas seems desirable. I've found them useful in enums and arrays already, and expect switch expressions will also benefit.
The examples that jumps to my mind in the Roslyn codebase are `WellKnownTypes` and `WellKnownMembers` (which are built from enums and arrays that we regularly expand).
```C#
enum Enum
{
    Alice,
    Bob,
}
```
```C#
new[]
{
   1,
   2,
}
```
```c#
_ = e switch
{
     Alice => "Alice",
     Bob => "Bob",
     // I expect to add entries here too
}
```

**Comment by theunrepentantgeek** [comment]
> Basically, no one is helped by disallowing this comma.

@CyrusNajmabadi, in this case I disagree with you.

_From a cognitive perspective ..._

When reading normal English prose (*), proper punctuation helps the brain process the text and parse its meaning. It's been shown that misspelled words slow readers by more than 10%; there's a similar slowdown caused by improper punctuation (but, unfortunately, I can't remember the figure).

(*) I suspect that this applies to other languages as well, but the studies I've read have focused on English.

I have to believe that the same would apply when reading code. Seeing a `,` is a signal to the reader that there's another clause following. When it's missing, that's a cognitive speed-bump that will slow down the reader.

_From the principle of least surprise ..._

Other languages permit trailing commas in many many places and they result in some very odd situations.

A friend shared this one from JavaScript:

```
> [1, 2, 3].length
3
> [1, 2, 3, ].length
3
> [1, 2, 3, , ].length
4
```

In any sane world, those arrays would have lengths 3, 4 and 5.

**Comment by alrz** [comment]
You can totally do `new[]{ 1, 2, }` and C# doesn't allow more than one trailing comma which is insignificant.

**Comment by sharwell** [comment]
Also note that it's easy to add an analyzer to forbid the `,` even if the language allows it, but impossible to add an analyzer that allows the `,` if the language doesn't.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi, in this case I disagree with you.

You are taking my quote out of context.  I specifically called out cases where there was semantic ambiguity.  And arrays, type-arg lists, and tuples would definitely be cases like that.  There is no such issue in the case being discussed here, and my statement was in the context of htat discussion.

**Comment by CyrusNajmabadi** [comment]
> I have to believe that the same would apply when reading code. Seeing a , is a signal to the reader that there's another clause following. When it's missing, that's a cognitive speed-bump that will slow down the reader.

This does not appear to be at all a problem given that hte language already allows for this in many effective cases.  Moreso, there's so little of a problem that no one has ever even asked for a way to prevent these pesky commas from being allowed in their codebase.  And, working in languages that require this, and codebases that lint-require this, i've never heard a single complaint from anyone.  But i have heard from many the benefit that the consistency and simplicity that things brings.

**Comment by stuartstein777** [comment]
Can't people who don't want unecessary trailing commas just set up style cop rules to squiggly them as errors. (I'm one of those people). THen people that want them can have them?

WHat's the issue?

**Comment by jnm2** [comment]
I'm having some cognitive dissonance. The look still (irrationally?) bothers me, but the benefit for line-based source control is so practical that I'm starting to hope StyleCop Analyzers allows me to enforce trailing commas everywhere they are allowed in the language so that I don't forget before it's too late.

**Comment by CyrusNajmabadi** [comment]
> StyleCop Analyzers allows me to enforce trailing commas everywhere they are allowed in the language so that I don't forget a trailing comma before it's too late.

This is how our linters work on the projects i work on.  I very much like it.

**Comment by yaakov-h** [comment]
> I'm starting to hope StyleCop Analyzers allows me to enforce trailing commas everywhere they are allowed in the language so that I don't forget a trailing comma before it's too late.

I believe it already does. 😄

**Comment by jnm2** [comment]
Looking sloppy is not the same thing as less readable. You called the amount of work to edit the commas trivial, but I can't understand how the impact on reading is not even more trivial.

Being bothered by something doesn't have to slow you down at all. If it does slow you down, I'm sure you can overcome that even more quickly than it'll take to absorb the new switch syntax (for example).

**Comment by qrli** [comment]
I think we all are used to `;` at end of each line, while it looks sloppy as English. 
Pascal gets that right by putting a `.` at end of program, but nothing gained.
Why is `,` so different?

**Comment by CyrusNajmabadi** [comment]
> The issue is that people seem to fall into one of two camps on this topic:
>    Readability is king. Trailing commas harm readability so they shouldn't be supported,

That doesn't even make sense to me.  Trailing commas simply makes each item consistent.  The list is simple `item <comma>` some number of times.  It's more readable to me to have each entry look similar to every other one, versus having the very last one be different.

**Comment by CyrusNajmabadi** [comment]
> Trailing commas make code less readable. They look messy and make the author look sloppy in my view. So as a fan of readable code, I oppose their use, even by others, as I may have to read their code one day.

"Non trailing commas make code less readable.  They look messy and make the author look sloppy in my view.  So, as a fan of readable code, I mandate their use, even by others, as I may have to read their code one day."

--

As mentioned above, the above isn't even untrue (though i wrote it to point out the inherent personally-opinionated nature of your post).  On my team our linting rules *mandate* this precisely so that code doesn't look 'sloppy' by virtue of the inconsistency introduced by their absence.  The major difference here is that i recognize that this is stylistic preference and so each form can be desired sensible by different groups. On the other hand, you seem to be insisting that only your stylistic preference can be accepted as the 'one true way', even if that's inconsistent with the rest of the language.

**Comment by Korporal** [comment]
If one cannot assume that a comma is always a separator but might also be a terminator depending on the context then IMHO that clutters the language and makes it harder to internalize the syntax. There is no right or wrong, the rules of the grammar are (so long as they're self consistent) up to the designers but supporting all of this mixed up and inconsistent rules for such basic tokens can only make the language harder to use in the end.

I also cannot think of many things that are more trifling than this, and I've been told here by experts that a feature must offer significant utility, this clearly does no such thing.

**Comment by YairHalberstadt** [comment]
@Korporal
The comma is not a terminator. It is a separator.
In some cases we allow a trailing seperator, and in some we don't, but they **must always be followed by a terminator**.

This isn't adding a new feature to the language. It's slightly relaxing the grammar of the language, and is far smaller a change to the compiler than plenty of the changes that happen on a weekly basis. As such the bar for entry is a little lower than new features.

**Comment by CyrusNajmabadi** [comment]
> I also cannot think of many things that are more trifling than this, and I've been told here by experts that a feature must offer significant utility, this clearly does no such thing.

That's not what experts here have said.  The position is that a feature must be worth it's cost.  The cost here is a pittance *given that the feature is already championed*.  The benefits are well worth it nd have been discussed at length in the thread.

> The only possible defence for trailing commas is making it easier to write code that generates code

That's not true at all.

> but supporting all of this mixed up and inconsistent rules 

Supporting this makes things *more* consistent across the language.

>  can only make the language harder to use in the end.

Woudln't the language be harder to use the *less* consistent it is?

**Comment by Neme12** [comment]
If I were to design a language there would be no separators at all. Having separators instead of terminators only creates problems. It makes the last line inconsistent and it makes editing harder. Just try using Pascal where the semicolon is a separator as opposed to a terminator.

Not only that, but separators also often introduce ambiguities. Why? With terminators, if you have N items, you have N terminators. With separators you have N - 1 separators unless you have 0 items, in which case you have 0 separators too. If you leave out the items themselves, it's impossible to distinguish whether there would have been 0 or 1 item in that place. Also, separators are *the* reason that we cannot have 1-element tuples.

If you're arguing that `,` is always a separator in English and therefore it makes code less readable (even though code isn't english), what about semicolons? They're separators in English too, yet we've all gotten used to using them as terminators in programming languages because separators are just so much inconvenient.

**Comment by YairHalberstadt** [comment]
@Neme12
I'm not sure I agree.

By using seperators you increase the information available to the compiler. Without seperators error detection is impossible as all inputs are valid. With seperators, leaving out something can be detected, which is useful to the programmer.

It's an old problem in information theory. The more dense your data, the harder it is to detect and recover from errors.

**Comment by DavidArno** [comment]
> If you're arguing that `,` is always a separator in English and therefore it makes code less readable (even though code isn't english), what about semicolons? 

Semicolons on the last statement are mandatory. Commas on the last item of a list are sometimes optional and sometimes a syntax error. In the first case, one just gets used to them. They are annoying, but they are consistent.

If commas were mandatory after the last item, then the same would apply to them as to semicolons. They aren't though, so the two are orthogonal.

**Comment by Neme12** [comment]
@YairHalberstadt Can you please clarify this with an example? What do you mean by "all inputs are valid"?

**Comment by gafter** [comment]
@Neme12 
> If I were to design a language, there would be no separators at all. Having separators instead of terminators only creates problems.

OK, so what would a method/function call look like?  Do you have no punctuation at all, or a trailing comma in the argument list?

**Comment by Neme12** [comment]
Either trailing commas (which I admit would take some getting used to), or there would only be semicolons instead of commas in the language.

**Comment by YairHalberstadt** [comment]
@neme12

I'm struggling to find a really decent example, so for now I'll retract until a come across one.

**Comment by CyrusNajmabadi** [comment]
> If I were to design a language, there would be no separators at all. Having separators instead of terminators only creates problems. It makes the last line inconsistent and it makes editing harder. Just try using Pascal where the semicolon is a separator as opposed to a terminator.

Every language designer eventually designs there way to the [one true language](https://en.wikipedia.org/wiki/Lisp_(programming_language)).

**Comment by Korporal** [comment]
> If I were to design a language, there would be no separators at all. Having separators instead of terminators only creates problems. It makes the last line inconsistent and it makes editing harder. Just try using Pascal where the semicolon is a separator as opposed to a terminator.
> 
> Not only that, but separators also often introduce ambiguities. Why? With terminators, if you have N items, you have N terminators. With separators you have N - 1 separators unless you have 0 items, in which case you have 0 separators too (it's not a 1 to 1 mapping). If you leave out the items themselves, it's impossible to tell whether there would have been 0 or 1 item in that place. Also, separators are _the_ reason that we cannot have 1-element tuples.
> 
> If you're arguing that `,` is always a separator in English and therefore it makes code less readable (even though code isn't english), what about semicolons? They're separators in English too, yet we've all gotten used to using them as terminators in programming languages because that's just so much more convenient.

I don't see how you can even define a grammar if there is no way to distinguish tokens from on another, even a space is a separator.

F# separates arguments in method calls (mostly) by simply spaces but it does need some way to do that.

**Comment by YairHalberstadt** [comment]
A space is not a seperator from the parsers perspective. 

The tokenizer strips out all spaces when creating tokens.

Then the parser treats some of those tokens (commas) as seperators.

**Comment by Korporal** [comment]
Why not alter the language to treat a trailing comma as legal here but issue a warning? Then the user can - if they so choose - either tolerate and ignore the warning or not?

**Comment by YairHalberstadt** [comment]
Because warnings are only issued for behaviour that can't be made illegal for whatever reason but could potentially be dangerous

**Comment by Korporal** [comment]
> A space is not a seperator from the parsers perspective.
> 

True a space is not a token as a comma is I agree but is still used to separate tokens as in here:

`public class some_identifer_1234`

> The tokenizer strips out all spaces when creating tokens.

Yes but it's the presence of the spaces that enables it to do that, that's all I'm saying here.
> 
> Then the parser treats some of those tokens (commas) as seperators.

**Comment by Neme12** [comment]
> even a space is a separator.

You can have trailing space after the last item. By separator I meant anything that disallows having a trailing separator. Anyway, my comment was out of topic. This issue is about C#.

**Comment by CyrusNajmabadi** [comment]
> Why not alter the language to treat a trailing comma as legal here but issue a warning? Then the user can - if they so choose - either tolerate and ignore the warning or not?


1. because warnings are effectively errors.
2. because warnings should be offered when there is something valid to warn about.  i.e. something that is *very very very very* likely a problem.  That's not hte case here.  There is nothing wrong here.  THat's the point.  It's a very reasonable coding pattern that we want to strictly *allow* and *support*.

That said, if you don't like this you are *totally capable of* writing your own analyzer to detect and prevent using this.  This is super common in the industry and people often write these sorts of 'linters'.  I would not be surprised if StyleCopAnalyzers adds support for this, likely with options to require this for people that want it, or disallow it for people that don't.  The language will support this being optional (like it does in other places), and users can decide which approach they want to take.

**Comment by CyrusNajmabadi** [comment]
The hypothetical explorations of languages that woudl not use separators doesn't seem to really be buying much here.  C# has separators.  They're not going anywhere.   This discussion is about what exceptionally tiny tweaks the grammar should allow in order to make the coding/editing experience nicer :)  It's def *not* about large scale discussions about whether or not separators should even exist :)

**Comment by Korporal** [comment]
It seems that this entire thread is discussing a change to a language feature that is not even released yet. That wasn't obvious to me.

**Comment by gafter** [comment]
Fixed in https://github.com/dotnet/roslyn/pull/32432 but still needs to be fixed in the specification.

**Comment by gafter** [comment]
Fixed in #2171
